junit5-engine/src/main/java/org/junit/gen5/engine/junit5/descriptor/JUnit5TestableFactory.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v1.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.junit.gen5.engine.junit5.descriptor;
+
+import static org.junit.gen5.commons.util.ReflectionUtils.loadClass;
+
+import java.lang.reflect.AnnotatedElement;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.junit.gen5.commons.util.Preconditions;
+import org.junit.gen5.commons.util.ReflectionUtils;
+
+public class JUnit5TestableFactory {
+
+	private static final String SEPARATORS = ":$#";
+
+	public JUnit5Testable fromUniqueId(String uniqueId, String engineId) {
+		Preconditions.notEmpty(uniqueId, "uniqueId must not be empty");
+		List<String> parts = split(uniqueId);
+		Preconditions.condition(parts.remove(0).equals(engineId), "uniqueId must start with engineId");
```

Comment:
If you are looking for a tie breaker...

I haven't looked at all the code, but I am leaning towards the position that the launcher should add the engine ID to the unique ID generated by the engine (or in an object that encapsulates the engine-generated ID and the engine ID). Another option would be that some code external to the engine (possibly the launcher) checks that the IDs generated by the engine are prefixed with the engine ID, but that leaks implementation details into the engine (why should the engine care how we make sure that IDs generated by multiple engines never conflict, or be forced to use a naming convention?)

Either way, if we want third-parties to add engines, then the engine ID would need to be a full class name, so the full IDs will be quite long. We should discuss how to make sure that most output that contains test IDs are easy for the user to parse (at least for the majority of users who never have a single test run containing tests from more that two engines). If there is a pass where all tests are enemerated before they are run, JUnit could see if there are conflicts ignoring engine ID, and if not only use the engine-generated ID in any stack traces. XML output would of course need either the full ID or the pieces necessary to construct it.